### PR TITLE
CP-53314: Read and watch <domain>/data/service in xenstore to DB

### PR DESCRIPTION
**Pull request by changlei-li**
The field `VM_guest_metrics.services` added previously (#6309) is used to store VM service information reported by guest tools.
Xapi does not use this data but just holds it for outside consumers.
This PR reads the xenstore path `data/service` and converts to `[("service_name/key", "value")]` pair list and then store to VM_guest_metrics.services.
The path to read is like:
`/local/domain/<id>/data/service/<service_name>/<key> = <value>`

### CODE REVIEW

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 1]
```diff
@@ -2710,6 +2710,72 @@ module VM = struct
         Domain.soft_reset ~xc ~xs di.Xenctrl.domid
     )
 
+  module Limit : sig
```
Comment (psafont) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2713]: Please add docc-comment here explaining the module `(** *)`

File: ocaml/xapi/xapi_guest_agent.ml [hunk 2]
```diff
@@ -236,6 +236,31 @@ let dead_domains : IntSet.t ref = ref IntSet.empty
...
+      List.fold_left
+        (fun acc key ->
+          let full_path_key = sub_path // key in
+          let db_key = service // key in
+          let value = lookup full_path_key in
+          (db_key, Option.value ~default:"" value) :: acc
+        )
+        acc (list sub_path)
+    )
+    [] services
```
Comment (psafont) [inline ocaml/xapi/xapi_guest_agent.ml:262]: I think this can be expressed with concat_map, and won't reverse the list:
```suggestion
  services
  |> List.concat_map (fun service ->
    let sub_path = base_path // service in
    list sub_path
    |> List.map (fun key ->
       let full_path_key = sub_path // key in
       let db_key = service // key in
       let value = lookup full_path_key in
       (db_key, Option.value ~default:"" value)
    )
  )
```

**Review by psafont — APPROVED** [review]
There's a new module that seems security-related, but its usage is not obvious to me, it lacks dopcumentation, and is missing unit-tests. Please at least add documentation and intended usage so we can change the module to make usre there aren't any footguns lurking around.

**Review by psafont — CHANGES_REQUESTED** [review]
There's a new module that seems security-related, but its usage is not obvious to me, it lacks dopcumentation, and is missing unit-tests. Please at least add documentation and intended usage so we can change the module to make usre there aren't any footguns lurking around. After that, unit-test will better document how touse it and highlight edge-cases.

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 3]
```diff
@@ -2813,58 +2880,87 @@ module VM = struct
...
+                      (quota, limit, acc)
                 in
                 let depth = depth - 1 in
                 List.fold_left
                   (ls_lR ~excludes ~depth root)
-                  (quota, acc) subdirs
+                  (quota, limit, acc) subdirs
+            in
+            let read_xs_path ?(excludes = StringSet.empty)
+                ?(limit = Limit.no_limit) ?(depth = 512) root (quota, result)
```
Comment (lindig) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2913]: If depth refers to a recursion/nesting limit - this seems to me too high. What hierarchy would be deeper than 16 levels?

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 4]
```diff
@@ -2813,58 +2880,87 @@ module VM = struct
...
                 let value_opt, subdirs = ls_l ~depth root dir in
-                let quota, acc =
+                let quota, limit, acc =
                   match value_opt with
                   | Some ((k, v) as entry) ->
-                      ( quota
-                        - Xenops_utils.xenstore_encoded_entry_size_bytes k v
-                      , entry :: acc
-                      )
+                      if
```
Comment (lindig) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2892]: This code needs a comment that explains its intent. It has something to do with maintaining a quota but it would be good to spell it out.

**Review by lindig — APPROVED** [review]
This needs documentation. I'm approving this but expect this to be updated.

File: ocaml/xapi/xapi_guest_agent.ml [hunk 5]
```diff
@@ -236,6 +236,31 @@ let dead_domains : IntSet.t ref = ref IntSet.empty
...
 
+(* Parse data/service which has the following structure:
+   data/service/<service_name>/<key> = <value>
+   data/service/<service_name>/<key> = <value>
+   ...
+   data/service/<service_name>/<key> = <value>
+   Read and convert to [(<service_name>/<key>, <value>)] pair list.
+   The list is intended to store in VM_guest_metrics.services at last *)
+let get_guest_services (lookup : string -> string option)
+    (list : string -> string list) =
```
[the remark this answers is not in the document]
Comment (changlei-li) [reply ocaml/xapi/xapi_guest_agent.ml:247]: `list` is defined in `xapi_xenops.ml` `update_vm`. **It lists all subdirs and keys in one depth under the given dir.** In its definition, the data source is a big string string map like `[("a", ""); ("a/b", ""); ("a/b/k1", "v1"); ("a/b/k2", "v2"); ("a/c/k1", "v1"); ("a/c/k2", "v2")]`. Then function `list "a"` will give all subdirs `["b", "c"]`. It needs to be noticed that the data source is not exposed to the caller in `xapi_guest_agent.ml`

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 3 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2913]: Yes, the depth is the dir path depth in xenstore. I think the default value is high too. I just follow the original number here. I don't want to change it now in this ticket. Every `guest_agent` path give a depth param, but when get `xsdata_state`, the depth is not given, and the default value is used.

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 4 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2892]: quota is like a global limit here. It's not the same with the limit I added in the PR. It deserves to consider it. Now, the function `get_state` is too long.

File: ocaml/xapi/xapi_guest_agent.ml (hunk 2 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_guest_agent.ml:262]: Thank you. I will try it.

**Comment by andyhhp** [comment]
xenstore is already a bottlekneck.  Xapi watching these keys really means xenopsd watching it and more traffic over message switch, which has proved to be an amplification vector in the past.

How many new watches does this create?

**Comment by lindig** [comment]
I would not expect these values to be updated frequently over the lifetime of a VM - so would not expect an impact from watches. @last-genius

**Comment by changlei-li** [comment]
Yes. Only single digit number of service-install data are written in data/service by guest agent. So, the path will be only written once after VM starting and expected not changed later.

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 6]
```diff
@@ -2710,6 +2710,101 @@ module VM = struct
...
     )
 
+  module Limit : sig
+    (** Set the limit when walking through a xenstore path. *)
+
+    (** Type representing the limit configuration. *)
+    type t
+
+    val make :
+         ?amount:int option
```
Comment (lindig) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2720]: Given that all limits are numerical. it might have been possible to represent no limit as `max_int` and avoid the handling of options. But options are fine, too.

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 7]
```diff
@@ -2710,6 +2710,101 @@ module VM = struct
...
+    (** Mark the limit as reached.
+      @param t The limit configuration.
+      @return A new limit configuration marked as reached. *)
+
+    val is_reached : t -> bool
+    (** Check if the limit has been reached.
+      @param t The limit configuration.
+      @return [true] if the limit has been reached, [false] otherwise. *)
+
+    val no_limit : t
```
Comment (lindig) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2763]: Does it make sense to define a default limit value, i.e. sensible limits?

**Review by lindig — APPROVED** [review]

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 1 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2713]: done

File: ocaml/xapi/xapi_guest_agent.ml (hunk 2 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_guest_agent.ml]: done

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 6 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2720]: I prefer option

**Comment by changlei-li** [comment]
It's strange the CI report while I passed the unit test in local environment.
> File "ocaml/xenopsd/test/test_xc.ml", line 15, characters 5-19:
15 | open Test_highlevel
          ^^^^^^^^^^^^^^
Error: Unbound module Test_highlevel

It should be the same with `test_cpuid.ml`. I will check it tomorrow.

File: ocaml/tests/test_guest_agent.ml [hunk 8]
```diff
@@ -468,9 +468,87 @@ module Initial_guest_metrics = Generic.MakeStateless (struct
...
+    in
+    let results =
+      List.filter_map
+        (fun (path, _) ->
+          if String.starts_with ~prefix:dir path then
+            let rest =
+              String.sub path (String.length dir)
+                (String.length path - String.length dir)
+            in
+            match List.filter (fun x -> x <> "") (String.split '/' rest) with
```
Comment (psafont) [inline ocaml/tests/test_guest_agent.ml:501]: Please drop Xstringext and instead use Astring:
```suggestion
            match Astring.fields ~empty:false rest with
```
https://ocaml.org/p/astring/latest/doc/Astring/String/index.html#val-fields

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 9]
```diff
@@ -2711,6 +2711,9 @@ module VM = struct
     )
 
   module Limit : sig
+    (** Set the limit when walking through a xenstore path. *)
```
Comment (psafont) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2714]: This comment should explain what is a "limit".

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 10]
```diff
@@ -2710,6 +2710,101 @@ module VM = struct
...
+      @param t The limit configuration.
+      @return [true] if the value string length is valid, [false] otherwise. *)
+
+    val mark_reached : t -> t
+    (** Mark the limit as reached.
+      @param t The limit configuration.
+      @return A new limit configuration marked as reached. *)
+
+    val is_reached : t -> bool
+    (** Check if the limit has been reached.
```
Comment (psafont) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2759]: This looks like a fuse to me. i.e. once the limit has been reached, it can't be changed back. Please document the property in here

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 11]
```diff
@@ -2710,6 +2710,101 @@ module VM = struct
...
+    val valid_value_length : string -> t -> bool
+    (** Check if the value string length is valid.
+      @param value The value string to check.
+      @param t The limit configuration.
+      @return [true] if the value string length is valid, [false] otherwise. *)
+
+    val mark_reached : t -> t
+    (** Mark the limit as reached.
+      @param t The limit configuration.
+      @return A new limit configuration marked as reached. *)
```
Comment (psafont) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2756]: This function should be removed, IMO, it should not be able to be used separately from `decrease_amount`.

File: ocaml/tests/test_guest_agent.ml (hunk 8 above)
Comment (changlei-li) [reply ocaml/tests/test_guest_agent.ml]: done

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 9 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml]: done

File: ocaml/xenopsd/xc/xenops_server_xen.ml (hunk 10 above)
Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2762]: done

File: ocaml/tests/test_guest_agent.ml (hunk 8 above)
Comment (psafont) [reply ocaml/tests/test_guest_agent.ml]: I'm trying to remove xstringext, it tends to be very inefficient :)

File: ocaml/xenopsd/xc/xenops_server_xen.ml [hunk 12]
```diff
@@ -2710,11 +2710,110 @@ module VM = struct
...
+    let is_reached t = t.reached
+
+    let no_limit = make ()
+  end
+
   let get_state vm =
     let uuid = uuid_of_vm vm in
     let vme = vm.Vm.id |> DB.read in
     (* may not exist *)
     let map_tr f l = List.rev_map f l |> List.rev in
```
Comment (contificate) [inline ocaml/xenopsd/xc/xenops_server_xen.ml:2815]: Unrelated to your changes, but this caught eye: probably not desirable as an idiom any more as I expect our version of OCaml has conversion into destination passing style (the "tail modulo cons" transformation).

Comment (changlei-li) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2815]: > 14 standard library functions made tail-recursive with Tail-Recursion-Modulo-Cons (TRMC), such as List.append and List.map

From 5.1.0

Comment (contificate) [reply ocaml/xenopsd/xc/xenops_server_xen.ml:2815]: Ah, unfortunate.

**Review by psafont — APPROVED** [review]

**Comment by changlei-li** [comment]
Xenopsd already has a quota limit mechanism for guest_agent when reading xenstore. Remove the amout, key_len, value_len limit in this PR. It's better to use the general quota limit mechanism to keep it simple.

File: ocaml/tests/test_guest_agent.ml [hunk 13]
```diff
@@ -468,9 +468,87 @@ module Initial_guest_metrics = Generic.MakeStateless (struct
...
+
+    let string_of_output_t = Test_printers.(assoc_list string string)
+  end
+
+  (* lookup and list are from Xapi_xenops.update_vm *)
+  let lookup state key = List.assoc_opt key state
+
+  let list state dir =
+    let dir =
+      if dir.[0] = '/' then
```
Comment (lindig) [inline ocaml/tests/test_guest_agent.ml:487]: This fails if `dir = ""`; don't know if that can happen but might want to have an assertion here.

File: ocaml/tests/test_guest_agent.ml [hunk 14]
```diff
@@ -468,9 +468,87 @@ module Initial_guest_metrics = Generic.MakeStateless (struct
...
+        String.sub dir 1 (String.length dir - 1)
+      else
+        dir
+    in
+    let results =
```
Comment (lindig) [inline ocaml/tests/test_guest_agent.ml:492]: This looks complicated. Would it not be easier to split a paths into a list of components and work on the resulting lists? The overall purpose of results is also not clear to me. A different name could help to convey what it is.

**Review by lindig — APPROVED** [review]

File: ocaml/tests/test_guest_agent.ml (hunk 13 above)
Comment (changlei-li) [reply ocaml/tests/test_guest_agent.ml]: Add a branch condition to avoid this.

File: ocaml/tests/test_guest_agent.ml (hunk 14 above)
Comment (changlei-li) [reply ocaml/tests/test_guest_agent.ml]: Change the name to list_subkeys, the steps are:
filter out the keys with given prefix
get the rest after prefix
split by `\` and get the first

File: ocaml/xapi/xapi_guest_agent.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/xapi/xapi_guest_agent.ml:247]:
```
utop # list_subkeys ["aaa/bbb/c", "x"] "aaa/bb";;
- : string list = ["b"]
```

Is this correct?

[the remark this answers is not in the document]
Comment (changlei-li) [reply ocaml/xapi/xapi_guest_agent.ml:247]: It's incorrect. I believe the real implementation of list in xapi_xenops.ml update_vm has this problem, too. Normally, it's a internal function and the caller only pass the parameter as a intact dir, so the problem is not exposed. It deserves to rasie a ticket to fix it.

[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/xapi/xapi_guest_agent.ml:247]:
```ocaml
  let ls assoc path =
    let split = String.split_on_char '/' in
    let prefix = split path in
    let paths = List.map (fun (path, _) -> split path) assoc in
    let rec entry prefix path =
      match prefix, path with
      | [], x::_ -> Some x
      | x::xs, y::ys when x = y -> entry xs ys
      | _ -> None
    in
      List.filter_map (entry prefix) paths
```
